fedimint-client-legacy/src/mint/mod.rs

```diff
@@ -994,4 +1100,109 @@ mod tests {
+            let select =
+                MintClient::select_notes_from_stream(stream, Amount::from_sats(multiplier * 1000))
+                    .await;
+            total_notes += select.unwrap().into_iter_items().count();
+        }
+        assert_eq!(total_notes / 100, 10);
+    }
+
+    #[test_log::test(tokio::test)]
+    async fn select_notes_returns_exact_amount_with_minimum_notes() {
+        let f = || {
+            sorted_note_stream(vec![
+                (Amount::from_sats(1), 10),
+                (Amount::from_sats(5), 10),
+                (Amount::from_sats(20), 10),
+            ])
+        };
+        assert_eq!(
+            MintClient::select_notes_from_stream(f(), Amount::from_sats(7))
+                .await
+                .unwrap(),
+            notes(vec![(Amount::from_sats(1), 2), (Amount::from_sats(5), 1)])
+        );
+        assert_eq!(
+            MintClient::select_notes_from_stream(f(), Amount::from_sats(20))
+                .await
+                .unwrap(),
+            notes(vec![(Amount::from_sats(20), 1)])
+        );
+    }
+
+    #[test_log::test(tokio::test)]
+    async fn select_notes_returns_next_smallest_amount_if_exact_change_cannot_be_made() {
+        let f = || {
+            sorted_note_stream(vec![
+                (Amount::from_sats(1), 1),
+                (Amount::from_sats(5), 5),
+                (Amount::from_sats(20), 5),
+            ])
+        };
+        assert_eq!(
+            MintClient::select_notes_from_stream(f(), Amount::from_sats(7))
+                .await
+                .unwrap(),
+            notes(vec![(Amount::from_sats(5), 2)])
+        );
+    }
+
+    #[test_log::test(tokio::test)]
+    async fn select_notes_uses_big_note_if_small_amounts_are_not_sufficient() {
+        let f = || {
+            sorted_note_stream(vec![
+                (Amount::from_sats(1), 3),
+                (Amount::from_sats(5), 3),
+                (Amount::from_sats(20), 2),
+            ])
+        };
+        assert_eq!(
+            MintClient::select_notes_from_stream(f(), Amount::from_sats(39))
+                .await
+                .unwrap(),
+            notes(vec![(Amount::from_sats(20), 2)])
+        );
+    }
+
+    #[test_log::test(tokio::test)]
+    async fn select_notes_returns_none_if_amount_is_too_large() {
+        let f = || sorted_note_stream(vec![(Amount::from_sats(10), 1)]);
+        assert!(matches!(
+            MintClient::select_notes_from_stream(f(), Amount::from_sats(100))
+                .await
+                .unwrap_err(),
+            MintClientError::InsufficientBalance(_, _)
+        ));
+    }
+
+    fn sorted_note_stream(
+        notes: Vec<(Amount, usize)>,
+    ) -> impl futures::Stream<Item = (Amount, usize)> {
+        futures::stream::iter(
+            notes
+                .into_iter()
+                .flat_map(|(amount, number)| vec![(amount, 0usize); number])
+                .sorted()
+                .rev(),
```

Comment:
`.rev` confuses me here. Said it's sorted... I didn't expect it to be in reversed order.

Reply:
Renamed